nkms_eth/actors.py

```diff
@@ -0,0 +1,263 @@
+            miner_ids.append(encoded_miner_id)
+
+        return tuple(miner_ids)
+
+    def eth_balance(self):
+        return self._blockchain._chain.web3.eth.getBalance(self.address)
+
+    def token_balance(self) -> int:
+        """Check miner's current token balance"""
+
+        # self._token_agent._check_contract_deployment()
+        balance = self._token_agent.call().balanceOf(self.address)
+
+        return balance
+
+    def withdraw(self, amount: int=0, entire_balance=False) -> str:
+        """Withdraw tokens"""
+
+        if entire_balance and amount:
+            raise Exception("Specify an amount or entire balance, not both")
+        elif entire_balance:
+            txhash = self.miner_agent.transact({'from': self.address}).withdrawAll()
+        else:
+            txhash = self.miner_agent.transact({'from': self.address}).withdraw(amount)
+
+        self._transactions.append(txhash)
+        self._blockchain._chain.wait.for_receipt(txhash, timeout=self._blockchain._timeout)
+
+        return txhash
+
+
+class PolicyAuthor(Actor):
+    """Alice"""
+
+    def __init__(self, address: bytes, policy_agent):
+        self.policy_agent = policy_agent
+        super().__init__(address)
+        self._arrangements = OrderedDict()    # Track authored policies by id
+
+    def make_arrangement(self, miner: Miner, periods: int, rate: int, arrangement_id: bytes=None) -> PolicyArrangement:
+        """
+        Create a new arrangement to carry out a blockchain policy for the specified rate and time.
+        """
+
+        value = rate * periods
+        arrangement = PolicyArrangement(author=self,
+                                        miner=miner,
+                                        value=value,
+                                        periods=periods)
+
+        self._arrangements[arrangement.id] = {arrangement_id: arrangement}
+        return arrangement
+
+    def get_arrangement(self, arrangement_id: bytes) -> PolicyArrangement:
+        """Fetch a published arrangement from the blockchain"""
+
+        blockchain_record = self.policy_agent.call().policies(arrangement_id)
+        author_address, miner_address, rate, start_block, end_block, downtime_index = blockchain_record
+
+        duration = end_block - start_block
+
+        miner = Miner(address=miner_address, miner_agent=self.policy_agent.miner_agent)
+        arrangement = PolicyArrangement(author=self, miner=miner, periods=duration)
+
+        arrangement.is_published = True
+        return arrangement
+
+    def revoke_arrangement(self, arrangement_id):
+        """Lookup the arrangement in the cache and revoke it on the blockchain"""
+        try:
+            arrangement = self._arrangements[arrangement_id]
+        except KeyError:
+            raise Exception('No such arrangement')  #TODO
+        else:
+            txhash = arrangement.revoke()
+        return txhash
+
+    def recruit(self, quantity: int) -> List[str]:
```

Comment:
A docstring would be nice. Without it, I can only figure out what it's doing by reading the code

Reply:
Yep, agreed - This needs a docstring, among other places.  `PolicyAuthor` is the centralization of "Alice logic". I'm trying to avoid colliding with the "sample" namespace here... I'm not tied to the name (or the entire method for that matter).